### PR TITLE
Use nmstate API for setting port vlan trunks in favor of using vlan-filtering script

### LINKED ISSUE #171: Support vlan-filtering configuration
Starting with version 0.3, nmstate allows users to configure [vlan-filtering options](https://github.com/nmstate/nmstate/pull/892), e.g.:

```yaml
interfaces:
  - name: linux-br0
    type: linux-bridge
    state: up
    bridge:
      port:
        - name: eth1
          vlan:
            mode: trunk
            trunk-tags:
            - id: 101
            - id-range:
                min: 500
                max: 599
            tag: 100
            enable-native: true
```

kubernetes-nmstate however handles the vlan-filtering configuration with a [workaround script](https://github.com/nmstate/kubernetes-nmstate/blob/master/build/bin/vlan-filtering) which makes sure to enable vlan-filtering by default and treat all bridge ports as trunks.

In order to introduce support for this feature while maintaining backwards compatibility, we should:

* Drop the custom script.
* In case no VLAN configuration is passed by the user, we should fill in defaults matching the behaviour of the script.

### Original description

Until we don't use NetworkManager to set vlan filtering (with or without nmstate) the linux bridge vlan_filtering flag is only set after desiredState is processed, so after reboot the flag is not going to be there until kubernetes-nmstate is not up and NodeNetworkState processed.

### PR DESCRIPTION
**Pull request by rhrazdil**
**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What this PR does / why we need it**:
This PR removes usage of vlan-filtering script which is hardcoded to set vlan trunk on every port of every bridge that is
added via NNCP desiredState.
Since nmstate now provides API for this configuration, there is no need for using the script.

The current behaviour is not changed for backwards compatibility.
When a NNCP adding a linux-bridge with state `up` is created, default vlan configuration is added to all bridge ports that do not define custom `vlan` configuration.

This at the same time allows users to define custom vlan configuration, we our script no longer interferes.

Closes #171 

**Special notes for your reviewer**:

```release-note
NONE 
```

### CODE REVIEW

**Review by phoracek — COMMENTED** [review]
Would you please introduce a new doc example with this new option? It is not a mandatory - feel free to keep this one focus on replacement of internals.

File: pkg/helper/bridges_test.go [hunk 1]
```diff
@@ -89,51 +140,54 @@ var _ = Describe("Network desired state bridge parser", func() {
...
-			Expect(obtainedBridgesAndPorts).To(HaveKeyWithValue("br1", BeEmpty()))
+			Expect(updatedDesiredState).To(MatchYAML(desiredState))
 		})
 	})
 	Context("when there are bridges up", func() {
 		BeforeEach(func() {
 			desiredState = someBridgesUp
 		})
-		It("should return the map with the bridges and ports", func() {
+		It("should add default vlan filtering to linux-bridge ports", func() {
```
Comment (AlonaKaplan) [inline pkg/helper/bridges_test.go:179]: I would add another test case for -
`When("some ports have vlan configuration while other don't"`...

File: test/e2e/handler/simple_bridge_and_bond_test.go [hunk 2]
```diff
@@ -118,9 +118,9 @@ func bondUpWithEth1Eth2AndVlan(bondName string) nmstate.State {
 
 var _ = Describe("NodeNetworkState", func() {
 	Context("when desiredState is configured", func() {
-		Context("with a linux bridge up with no ports", func() {
+		Context("with a linux bridge up with some ports", func() {
```
Comment (AlonaKaplan) [inline test/e2e/handler/simple_bridge_and_bond_test.go:121]: You wrote on the commit message -
```
linux-bridge CNI configures vlan_filtering flag for veth pairs, so there shouldn't be a need for it here.
```
Can you please explain what is vlan filtering on veth pair and what problem it solves?
I mean, what if the CNI didn't configure the vlan filtering on the veth, would it cause any problem?

**Comment by AlonaKaplan** [comment]
General question, why not using `NodeNetworkConfigurationPolicy mutating webhook` to mutate the policy with the default vlan filtering values instead of doing the same operation separately for each node?

File: pkg/helper/client.go [hunk 3]
```diff
@@ -126,6 +126,12 @@ func ApplyDesiredState(client client.Client, desiredState shared.State) (string,
 	// working fine after apply
 	probes := probe.Select(client)
 
+	desiredState, err := applyDefaultVlanFiltering(desiredState)
```
Comment (AlonaKaplan) [inline pkg/helper/client.go:129]: If not doing it in the webhook as suggested [here](https://github.com/nmstate/kubernetes-nmstate/pull/793#issuecomment-896573208), I would at least update the enactment as well. So it will be more obvious that the bridge default value are desired and there is no diff between the desired and the current state.

File: pkg/helper/bridges.go [hunk 4]
```diff
@@ -4,33 +4,58 @@ import (
...
+	ifaces := gjson.ParseBytes(result).Get("interfaces").Array()
+	for ifaceIndex, iface := range ifaces {
+		if !isLinuxBridgeUp(iface) {
+			continue
 		}
+		for portIndex, port := range iface.Get("bridge.port").Array() {
+			if hasVlanConfiguration(port) {
+				continue
+			}
+			result, err = sjson.SetBytes(result, fmt.Sprintf("interfaces.%d.bridge.port.%d.vlan", ifaceIndex, portIndex), defaultVlanFiltering)
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/helper/bridges.go:41]: the pvid is set to 1 by default when vlan is configured in nmstate, I have compared all the bridge properties with this patch and the current state to make sure that there is indeed no difference from current behaviour

File: test/e2e/handler/simple_bridge_and_bond_test.go (hunk 2 above)
Comment (rhrazdil) [reply test/e2e/handler/simple_bridge_and_bond_test.go:121]: Now when I think about this sentence, I see is as not relevant really, because this change doesn't affect it in any way.
But since you're asking, the way I understand it is that there are northbound and southbound bridge ports, northbound are the ports configured by nmstate, southbound are the veth pairs created by CNI that connect virt-launcher pod with the bridge.
These veth pairs need to have the vlan filtering configured in compliance to what we configure by nmstate (or the vlan_filtering script respectively) to be able to communicate

File: pkg/helper/client.go (hunk 3 above)
Comment (rhrazdil) [reply pkg/helper/client.go]: Done.
I have added a webhook that adds the default vlan configuration to the NNCP, so both NNCP and NNCEs have the desiredState with default values
I'm still adding the defaults in the Reconcile to make sure old NNCPs have the defaults after nmstate update.

We could also update the NNCP during the Reconcile, which I'm not doing in the current state of this PR, because I'm not convinced it's worth the API load this would trigger - IIUC, each node would reconciled each NNCP twice in cases when the NNCP would be updated right?

Comment (rhrazdil) [reply pkg/helper/client.go]: On second thought, given that the webhook should add the defaults in majority of cases, I suppose we can check the default in Reconcile and update the policy case it's needed

Comment (rhrazdil) [reply pkg/helper/client.go]: @AlonaKaplan 
After discussion with Petr, we decided it's not needed nor desired to include internal defaults into NNCP. The reason for this being that there are many various defaults set by the system as well.

The current change includes the defaults vlan setting into the NNCE to avoid the discrepancy between desiredState and NNS (since the 'default' we set for VLANs are not really system defaults).

File: pkg/helper/bridges_test.go (hunk 1 above)
Comment (rhrazdil) [reply pkg/helper/bridges_test.go:247]: Added

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 5]
```diff
@@ -155,6 +155,11 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
 		return ctrl.Result{}, err
 	}
 
+	err = r.setPolicyDesiredStateDefaults(instance)
```
Comment (AlonaKaplan) [inline controllers/nodenetworkconfigurationpolicy_controller.go:158]: Why are you updating the policy instance and not the enactment directly?
For example. you had to change the `incrementUnavailableNodeCount` behaviour because of the policy instance change, updating the enactment directly would avoid it.

File: pkg/webhook/nodenetworkconfigurationpolicy/nodenetworkconfigurationpolicy_suite_test.go [hunk 6]
```diff
@@ -17,7 +18,8 @@ import (
 
 func TestUnit(t *testing.T) {
 	RegisterFailHandler(Fail)
-	RunSpecs(t, "NNCP Webhook Test Suite")
+	junitReporter := reporters.NewJUnitReporter("junit.webhook-webhook_suite_test.xml")
+	RunSpecsWithDefaultAndCustomReporters(t, "NNCP Webhook Test Suite", []Reporter{junitReporter})
```
Comment (AlonaKaplan) [inline pkg/webhook/nodenetworkconfigurationpolicy/nodenetworkconfigurationpolicy_suite_test.go:22]: Why is this change needed?

File: test/e2e/handler/utils.go [hunk 7]
```diff
@@ -123,6 +124,12 @@ func updateDesiredStateAtNodeAndWait(node string, desiredState nmstate.State) {
 	waitForAvailableTestPolicy()
 }
 
+func desiredStateWithDefaults(desiredState nmstate.State) nmstate.State {
```
Comment (AlonaKaplan) [inline test/e2e/handler/utils.go:127]: Please create a manual hard coded version of the desired state.
The e2e tests shouldn't use the same code the dev is using for verification. Bug in the dev code won't be discovered this way.

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 5 above)
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:158]: I have considered two options here:
1. update the enactment + apply the defaults to the policy right before setting the desiredState in `pkg/helper/client.go ApplyDesiredState`
2. just update the NNCP, initialize the NNCE based on the updated version and use it to also set the desired state

I considered the 2. option to be more straightforward, because it doesn't mean repeating the same work - parsing the desiredState and applying defaults to it.
It did require to change the `incrementUnavailableNodeCount` so the the function doesn't modify the parameter passed to it, which I think is better anyway.

File: pkg/webhook/nodenetworkconfigurationpolicy/nodenetworkconfigurationpolicy_suite_test.go (hunk 6 above)
Comment (rhrazdil) [reply pkg/webhook/nodenetworkconfigurationpolicy/nodenetworkconfigurationpolicy_suite_test.go:22]: I noticed that the webhook tests didn't produce xml output, I'll remove this  change from this PR as it's not related anymore

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 5 above)
Comment (AlonaKaplan) [reply controllers/nodenetworkconfigurationpolicy_controller.go:158]: What about option 3 - use the desired state from the enactment on `ApplyDesiredState`. I believe it is more correct even not related to this PR. Since the enactment represent the desired state of the current node.ApplyDesiredState
And this way you won't have to carry about updating the policy instance.

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Yes, that makes sense as well. Updated accordingly

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 8]
```diff
@@ -160,6 +160,12 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
 		log.Error(err, "Error initializing enactment")
 	}
 
+	enactmentInstance, err := r.applyEnactmentDefaults(instance)
```
Comment (AlonaKaplan) [inline controllers/nodenetworkconfigurationpolicy_controller.go:163]: Not related to this PR, but what is the reason for not returning an error when `initializeEnactment` fails.

Comment (AlonaKaplan) [inline controllers/nodenetworkconfigurationpolicy_controller.go:163]: I would move this code to be inside `initializeEnactment` so it will be created with the default from the beginning and there won't be a need for 2 extra api calls (retrieving and updating the enactment).

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:163]: I'm not sure if moving this code into `initEnactment` is worth saving the calls, as it would make the function a little too complex IMO.
Mainly because `inititalizeEnactment` returns conditions of the 
WDYT about a compromise - saving 1 API call - by setting the desiredState with defaults in the `initializeEnactment` function and instead of this code only fetch the updated enactment?
That saves one API (write) call.

The main issue I have with moving this whole piece to `initializeEnactment` is that the function returns previous enactment  Status, and I don't like the idea of mixing the previous enactmentStatus with the current one.

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:163]: Actually, I think it is related now.
Previously, I suppose, we were okay with the enactment not created, as it didn't have any consequence for applying the policy desiredState.
But the current code passes the desiredState in the enactment, so if it happened to be nil, it would blow.

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: addressed in separate commit for easier review https://github.com/nmstate/kubernetes-nmstate/pull/793/commits/c1ec4a4fb04ba0f8c95825112b40c053a032b5c8

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: [Updated](https://github.com/nmstate/kubernetes-nmstate/pull/793/commits/c1ec4a4fb04ba0f8c95825112b40c053a032b5c8)

File: docs/examples/linux-bridge-vlan.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
+apiVersion: nmstate.io/v1beta1
```
Comment (phoracek) [inline docs/examples/linux-bridge-vlan.yaml:1]: Could we add it to the list in the README?

File: docs/examples/linux-bridge-vlan.yaml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+            - name: eth1
+            vlan:
+              mode: trunk
+              trunk-tags:
+                - id: 101
+                - id-range:
+                    max: 500
+                    min: 200
+              tag: 100
+              enable-native: true
```
Comment (phoracek) [inline docs/examples/linux-bridge-vlan.yaml:29]: Please end it with a newline

File: docs/examples/linux-bridge-vlan.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
+apiVersion: nmstate.io/v1beta1
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: linux-bridge-vlan
+spec:
+  desiredState:
+    interfaces:
+      - name: br1
+        description: Linux bridge with eth1 as a port with vlan configuration
```
Comment (phoracek) [inline docs/examples/linux-bridge-vlan.yaml:9]: custom vlan configuration?

**Review by phoracek — CHANGES_REQUESTED** [review]
Minor nits.

Could we also test that we are able to disable vlan-filtering (maybe with vlan: {})?

File: docs/examples/linux-bridge-vlan.yaml (hunk 9 above)
Comment (rhrazdil) [reply docs/examples/linux-bridge-vlan.yaml:1]: Done

File: docs/examples/linux-bridge-vlan.yaml (hunk 11 above)
Comment (rhrazdil) [reply docs/examples/linux-bridge-vlan.yaml]: Done

File: docs/examples/linux-bridge-vlan.yaml (hunk 10 above)
Comment (rhrazdil) [reply docs/examples/linux-bridge-vlan.yaml]: Done

**Comment by rhrazdil** [comment]
>Could we also test that we are able to disable vlan-filtering (maybe with vlan: {})?


Good point, added this case to e2e (https://github.com/nmstate/kubernetes-nmstate/pull/793/commits/bc345316695f1189e4c6fd9e6c731c19735dcb02#diff-f9955f6099eeab92950ef0ee96d13b0afd37827dbef68b7c9b92c80e9d6aabf6R139) and unit tests as well

**Review by phoracek — APPROVED** [review]
This is great stuff, kudos!

/lgtm
/approve
